devito/types/caching.py

```diff
@@ -0,0 +1,158 @@
+import gc
+import weakref
+
+import sympy
+
+__all__ = ['Cached', '_SymbolCache', 'CacheManager']
+
+_SymbolCache = {}
+"""The symbol cache."""
+
+
+class AugmentedWeakRef(weakref.ref):
+
+    def __new__(cls, obj, meta):
+        obj = super().__new__(cls, obj)
+        obj.nbytes = meta.get('nbytes', 0)
+        return obj
+
+
+class Cached(object):
+
+    """
+    Mixin class for cached symbolic objects.
+    """
+
+    @classmethod
+    def _cache_key(cls, *args, **kwargs):
+        """
+        A unique, deterministic cache key from the input arguments.
+
+        Notes
+        -----
+        To be implemented by subclasses.
+        """
+        raise NotImplementedError
+
+    @classmethod
+    def _cache_get(cls, key):
+        """
+        Retrieve the object corresponding to a given key. If the key is not in
+        the symbol cache or if the mapped object is not alive anymore, returns None.
+
+        Parameters
+        ----------
+        key : object
+            The cache key. It must be hashable.
+        """
+        if key in _SymbolCache:
+            # There is indeed an object mapped to `key`. But is it still alive?
+            obj = _SymbolCache[key]()
+            if obj is None:
+                # Cleanup _SymbolCache (though practically unnecessary)
+                del _SymbolCache[key]
+                return None
+            else:
+                return obj
+        else:
+            return None
+
+    def __init__(self, key):
+        """
+        Store `self` in the symbol cache.
+
+        Parameters
+        ----------
+        key : object
+            The cache key. It must be hashable.
+        """
+        # Precompute hash. This uniquely depends on the cache key
+        self._cache_key_hash = hash(key)
+
+        # Add ourselves to the symbol cache
+        _SymbolCache[key] = AugmentedWeakRef(self, self._cache_meta())
+
+    def __init_cached__(self, key):
+        """
+        Initialise `self` with a cached object state.
+
+        Parameters
+        ----------
+        key : object
+            The cache key of the object whose state is used to initialize `self`.
+            It must be hashable.
+        """
+        self.__dict__ = _SymbolCache[key]().__dict__
+
+    def __hash__(self):
+        """
+        The hash value of an object that caches on its type is the
+        hash value of the type itself.
+        """
+        return self._cache_key_hash
+
+    def _cache_meta(self):
+        """
+        Metadata attached when ``self`` is added to the symbol cache.
```

Comment:
'self' has being used with one " ' " above, like in `__init_cahed__` `__init__` etc